### ISSUE TITLE
Prometheus Metrics for reverse_proxy upstreams

### ISSUE THREAD

**Issue by genofire** [caddyserver/caddy#4140]
It would be nice to know, which Upstream give which __latency__, __statuscode__ or __health_status__.

Maybe it is also nice to enable it per __path__ (related issue #4016).

**Comment by francislavoie** [comment]
@uniuuu I think that's just showing how many requests reached the `reverse_proxy` handler. The question was about metrics for _which upstreams_ were hit, i.e. when you have more than one upstream configured for your `reverse_proxy`.

**Comment by lfkdev** [comment]
Still a much needed feature. Seeing which upstreams are used is a basic information for a loadbalancer and should be included in the metrics.

**Comment by mholt** [comment]
Might only be partially resolved by #4935, so reopening

**Comment by farhan-ct** [comment]
Hi @uniuuu 

Could you please explain how you did the /metrics setup? Because simply turning on the metrics api was not giving me any middleware metrics except the admin metrics.

Please check [https://github.com/caddyserver/caddy/issues/4309#issuecomment-1353007832](url)

I request you to help me out with this. Thank you!

**Comment by lfkdev** [comment]
@farhan-ct you can enable /metrics in the Caddyconfig and you'll see metrics like which endpoints are online and if they're in a healthy state. 
Example:
```
[..]
caddy_reverse_proxy_upstreams_healthy{upstream="123.123.123.123:1337"} 1
caddy_reverse_proxy_upstreams_healthy{upstream="123.123.124.123:1338"} 0
caddy_reverse_proxy_upstreams_healthy{upstream="123.123.125.123:1339"} 1
[..]
```

Make sure you use a new Caddy Version, that feature was just introduced a little while ago.
https://caddyserver.com/docs/metrics

**Comment by farhan-ct** [comment]
> @farhan-ct you can enable /metrics in the Caddyconfig and you'll see metrics like which endpoints are online and if they're in a healthy state. Example:
> 
> ```
> [..]
> caddy_reverse_proxy_upstreams_healthy{upstream="123.123.123.123:1337"} 1
> caddy_reverse_proxy_upstreams_healthy{upstream="123.123.124.123:1338"} 0
> caddy_reverse_proxy_upstreams_healthy{upstream="123.123.125.123:1339"} 1
> [..]
> ```
> 
> Make sure you use a new Caddy Version, that feature was just introduced a little while ago. https://caddyserver.com/docs/metrics

Yes I enabled that and am able to see the metrics that you just mentioned. However, you can find in this link [[https://github.com/caddyserver/caddy/issues/4309#issuecomment-1353007832](https://github.com/caddyserver/caddy/issues/url)](url)  that the @rugk is able to get some other metrics starting with caddy_http_..., request and response stuff. I am struggling on how to get those.

Also, based this PR [https://github.com/caddyserver/caddy/pull/3709](url)  You can see the sample code in the introduction that's providing the caddy_http_request* metrics which am unable to find in my v2.6.2 (latest) version

**Comment by hairyhenderson** [comment]
@farhan-ct please read the 2.6 release notes - metrics need to be explicitly enabled now: https://github.com/caddyserver/caddy/releases/tag/v2.6.0

**Comment by DBarney** [comment]
I realize that it has been a few years since this was opened. But it would be extremely good to be able to break down which upstream was used to generate the response.

I have a config with multiple `reverse_proxy` blocks, and unfortunately they all get lumped together in the metrics. Which makes them fairly useless as far as metrics go. There is no way to identify which group of upstreams is having issues or problems, and generate alerts on those.

I would even settle for being able to identify different `reverse_proxy` blocks with a specific label and tag. That way I could separate the metrics even a little bit.

**Comment by not-not-the-imp** [comment]
In response to @farhan-ct the only way I was able to get non admin metrics was to have a separate end point than the admin one.

Here is what I did.

## Enable Metrics at the top level
You need metrics enabled at the top level, as per the doc
```
{
        servers {
                metrics
        }
}
```
That will provide a small part of the metrics eg the admin one on `localhost:2019/metrics`
To test it 

## Create a separate subdomain:
To get the non admin metrics like `caddy_http_` I had to configure a subdomain 

### caddy version 2.6.2
The doc says `basic_auth`, but it only worked with `basicauth`
```
metrics.example.org {
        basicauth {
                # Username "Bob", password "hiccup"
		Bob $2a$14$Zkx19XLiW6VYouLHR5NmfOFU0z2GTNmpkT/5qqR7hx4IjWJPDhjvG
        }
        metrics
}
```

### Caddy version 2.6.8
```
metrics.example.org {
        basic_auth {
                # Username "Bob", password "hiccup"
		Bob $2a$14$Zkx19XLiW6VYouLHR5NmfOFU0z2GTNmpkT/5qqR7hx4IjWJPDhjvG
        }
        metrics
}
```

## the prometheus config

The caddy config will serve the metrics over HTTPS.
It will server the metrics on both `/` and `/metrics` ( I choose to scrape `/`)
I added a basic username/password authentication, so I provide it too
```
scrape_configs:
...
  - job_name: "caddy"
    scheme: https
    static_configs:
      - targets: ['metrics.example.org']
    metrics_path: "/"
    basic_auth:
        username: bob
        password: 'hiccup'
        }
        metrics
}
```
## Prometheus Dashboard
This [Caddy exporter dashboard](https://grafana.com/grafana/dashboards/14280-caddy-exporter/) then works fine for the most part
https://grafana.com/grafana/dashboards/14280-caddy-exporter/

## What available for upstreams
There is still no metrics available per upstreams, only an on/off status.
```
# HELP caddy_reverse_proxy_upstreams_healthy Health status of reverse proxy upstreams.
# TYPE caddy_reverse_proxy_upstreams_healthy gauge
caddy_reverse_proxy_upstreams_healthy{upstream="localhost:3000"} 1
caddy_reverse_proxy_upstreams_healthy{upstream="localhost:3001"} 1
caddy_reverse_proxy_upstreams_healthy{upstream="localhost:3002"} 1
```

## What I'd like to have for upstream.
* A dedicated label
* Every `duration_seconds_bucket` per upstream (both `request` and `response`)
* Every `size_bytes_bucket` per upstream (both `request` and `response`)

I'm not sure how feasible it is to have a dedicated label, but here is how I would see it work
```
gatus.example.org {
        # metrics_label default to FQDN with '.' replaced by '_'
        metrics_label  gatus_example_org
        reverse_proxy localhost:3000
}

Would a PR that does something like the above be considered ?

**Comment by ArcticLampyrid** [comment]
I want to record the status of the configured reverse proxy upstreams using Prometheus.
The information can be fetched via [Caddy API `/reverse_proxy/upstreams`](https://caddyserver.com/docs/api#get-reverse-proxyupstreams). Can we add them to `metrics`?

**Comment by mohammed90** [comment]
> I want to record the status of the configured reverse proxy upstreams using Prometheus.
> The information can be fetched via [Caddy API `/reverse_proxy/upstreams`](https://caddyserver.com/docs/api#get-reverse-proxyupstreams). Can we add them to `metrics`?

That's already available. They're only missed in 2.9 due to a bug in a refactor. They'll be included *again* in 2.10.

**Comment by mholt** [comment]
(2.10 is released btw)
